File: bindings/python/rrc/identity.py
# SPDX-License-Identifier: 0BSD

from __future__ import annotations

from . import ffi
from .errors import Error, map_code


class Identity:
    def __init__(self, handle: int) -> None:
        self._handle = int(handle)

    @classmethod
    def generate(cls) -> Identity:
        h = ffi.lib.rrc_identity_generate()
        if h == 0:
            raise Error(Error.INTERNAL)
        return cls(h)

    @classmethod
    def load(cls, path: str) -> Identity:
        h = ffi.lib.rrc_identity_load(path.encode("utf-8"))
        if h == 0:
            raise Error(Error.IO)
        return cls(h)

    def save(self, path: str) -> None:
        map_code(ffi.lib.rrc_identity_save(self._handle, path.encode("utf-8")))

    def hash_bytes(self) -> bytes:
        buf = bytearray(ffi.HASH_LEN)
        written = ffi.ctypes.c_size_t(0)
        arr = (ffi.ctypes.c_uint8 * ffi.HASH_LEN).from_buffer(buf)
        map_code(
            ffi.lib.rrc_identity_hash(
                self._handle,
                arr,
                ffi.HASH_LEN,
                ffi.ctypes.byref(written),
            )
        )
        return bytes(buf[: written.value])

    def seed_destination(self, dest_hash: bytes) -> None:
        if len(dest_hash) != ffi.HASH_LEN:
            raise Error(Error.INVALID_ARG)
        arr = (ffi.ctypes.c_uint8 * ffi.HASH_LEN).from_buffer_copy(dest_hash)
        map_code(ffi.lib.rrc_identity_seed_destination(self._handle, arr, ffi.HASH_LEN))

    @property
    def handle(self) -> int:
        return self._handle

    def close(self) -> None:
        if self._handle:
            ffi.lib.rrc_identity_destroy(self._handle)
            self._handle = 0

    def __enter__(self) -> Identity:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass

    def __repr__(self) -> str:
        return f"Identity(handle={self._handle})"

File: bindings/java/src/io/quad4/rrc/Identity.java
// SPDX-License-Identifier: 0BSD

package io.quad4.rrc;

import com.sun.jna.ptr.LongByReference;
import io.quad4.rrc.ffi.RrcLibrary;

public final class Identity implements AutoCloseable {
    private long handle;

    private Identity(long handle) {
        this.handle = handle;
    }

    public static Identity generate() {
        long h = RrcLibrary.INSTANCE.rrc_identity_generate();
        if (h == 0) {
            throw new RrcException(RrcException.INTERNAL);
        }
        return new Identity(h);
    }

    public static Identity load(String path) {
        long h = RrcLibrary.INSTANCE.rrc_identity_load(path);
        if (h == 0) {
            throw new RrcException(RrcException.IO);
        }
        return new Identity(h);
    }

    public void save(String path) {
        Rrc.check(RrcLibrary.INSTANCE.rrc_identity_save(handle, path));
    }

    public byte[] hashBytes() {
        byte[] buf = new byte[Rrc.HASH_LEN];
        LongByReference written = new LongByReference();
        Rrc.check(RrcLibrary.INSTANCE.rrc_identity_hash(handle, buf, buf.length, written));
        byte[] out = new byte[(int) written.getValue()];
        System.arraycopy(buf, 0, out, 0, out.length);
        return out;
    }

    public void seedDestination(byte[] destHash) {
        if (destHash.length != Rrc.HASH_LEN) {
            throw new RrcException(RrcException.INVALID_ARG);
        }
        Rrc.check(RrcLibrary.INSTANCE.rrc_identity_seed_destination(handle, destHash, destHash.length));
    }

    public long handle() {
        return handle;
    }

    @Override
    public void close() {
        if (handle != 0) {
            RrcLibrary.INSTANCE.rrc_identity_destroy(handle);
            handle = 0;
        }
    }
}
